### ISSUE TITLE
What is the scope of "Linting"

### ISSUE THREAD

**Issue by imciner2** [PlasmaFAIR/fortitude#255]
One thing that I think would be useful to pin down is, what is the scope of what we want this "linter" tool to accomplish, since I think sometimes there can be some differing opinions about exactly what goes into a linter.

For starters, I am assuming there is a goal to eventually include some static analysis parts? e.g., unused variables and the like?

What about parts that are against the standard (so flagged by compilers as errors), but might still be interesting - e.g., duplicate case values in a select statement?

**Comment by ZedThree** [comment]
My personal view is that if some warning is:

a) useful (to a decent fraction of the community -- not everything will be useful to everyone),
b) we can do it, and
c) we can do it fast

then we should probably do it. If something is very useful but also expensive, then maybe we relegate it to something that has to be explicitly turned on.

I'd definitely like to get some static analysis in, unused variables etc. This might be much easier if we're able to build on LFortran.

Similarly for syntax errors -- tree-sitter doesn't give great information about what's gone wrong, so that's another reason to investigate LFortran.

Non-portable features would be amazing to flag, but I'm a bit worried this implies having a parser which is a superset of all compilers' parsers.

**Comment by Beliavsky** [comment]
Perhaps functionality that is not already in gfortran should be prioritized. For example, it's good to flag unused variables, but gfortran already has a compiler flag to do this. I wonder if a compiler such as gfortran and Fortitude can even be combined:

1. compile the code with gfortran with warnings on
2. if there are syntax errors, the user must fix
3. if the code compiles but there are warnings, feed the error message to Fortitude along with the code
4. have Fortitude fix the code and goto 1
5. run Fortitude on the code that compiles cleanly and fix remaining problems if possible

I have written Python scripts that take gfortran warnings, for example about unused variables, and use them to fix code. I also wrote a small Python [agent](https://github.com/Beliavsky/ChatGPT-Fortran-generator) that has ChatGPT generate Fortran code and then fix it based on gfortran warnings. The agent could also be asked to fix flaws found by Fortitude.

**Comment by ZedThree** [comment]
Yes, agreed, we should prioritise warnings not in existing compilers -- although I believe most (all?) Fortran compilers do not offer automatic fixits, so that may be a good reason for us to include e.g. unused variables or imports.

Many IDEs can offer something like you suggest for compiler warnings. Certainly emacs can via fly-check/flymake, although again without automatic fixes.
